lib/utils/fakeconn.go

```diff
@@ -201,6 +198,64 @@ func (c *ChConn) SetWriteDeadline(t time.Time) error {
 	return nil
 }
 
+// CancelableChConn is a wrapped SSH channel connection that supports deadlines.
+type CancelableChConn struct {
+	// ChConn is the base wrapped SSH channel connection.
+	*ChConn
+	// reader is the part of the pipe that clients read from.
+	reader net.Conn
+	// writer is the part of the pipe that receives data from SSH channel.
+	writer net.Conn
+}
+
+// NewCancelableChConn returns a new instance of wrapped SSH channel connection
+// that supports deadlines.
+func NewCancelableChConn(conn ssh.Conn, ch ssh.Channel) *CancelableChConn {
+	reader, writer := net.Pipe()
+	c := &CancelableChConn{
+		ChConn: NewChConn(conn, ch),
+		reader: reader,
+		writer: writer,
+	}
+	// Start copying from the SSH channel to the writer part of the pipe. The
+	// clients are reading from the reader part of the pipe (see Read below).
+	//
+	// This goroutine stops when either the SSH channel closes or this
+	// connection is closed e.g. by a http.Server (see Close below).
+	go io.Copy(writer, ch)
+	return c
+}
+
+// Read reads from the channel.
+func (c *CancelableChConn) Read(data []byte) (int, error) {
+	return c.reader.Read(data)
+}
+
+// SetDeadline sets the channel connection read/write deadlines.
+func (c *CancelableChConn) SetDeadline(t time.Time) error {
+	return c.reader.SetDeadline(t)
```

Comment:
Which makes me wonder whether the original `ChConn` should return errors instead of `nil` for the APIs it does not implement - WDYT? Would be at least interesting to see whether it breaks anywhere in tests.

Reply:
Yeah, I'm pretty sure things will break in mysterious ways :) A lot of Go runtime uses deadlines and will trip if we start returning errors. I'm actually more surprised that things seems to work ok without deadlines implemented on ChConn (well, besides this websockets issue).